src/main/java/org/disq_bio/disq/HtsjdkReadsRddStorage.java

```diff
@@ -0,0 +1,189 @@
+   *
+   * @param path the file or directory to read from
+   * @return a {@link HtsjdkReadsRdd} that allows access to the reads
+   * @throws IOException if an IO error occurs while determining the format of the files and reading
+   *     the header
+   */
+  public HtsjdkReadsRdd read(String path) throws IOException {
+    return read(path, null);
+  }
+
+  /**
+   * Read reads from the given path, using the given traversal parameters to filter the reads. The
+   * input files may be in any format (BAM/CRAM/SAM).
+   *
+   * @param path the file or directory to read from
+   * @param traversalParameters parameters that determine which reads should be returned, allows
+   *     filtering by interval
+   * @param <T> the type of Locatable for specifying intervals
+   * @return a {@link HtsjdkReadsRdd} that allows access to the reads
+   * @throws IOException if an IO error occurs while determining the format of the files
+   */
+  public <T extends Locatable> HtsjdkReadsRdd read(
+      String path, HtsjdkReadsTraversalParameters<T> traversalParameters) throws IOException {
+
+    FileSystemWrapper fileSystemWrapper =
+        useNio ? new NioFileSystemWrapper() : new HadoopFileSystemWrapper();
+
+    String firstSamPath;
+    if (fileSystemWrapper.isDirectory(sparkContext.hadoopConfiguration(), path)) {
+      firstSamPath =
+          fileSystemWrapper.firstFileInDirectory(sparkContext.hadoopConfiguration(), path);
+    } else {
+      firstSamPath = path;
+    }
+    SamFormat samFormat = SamFormat.fromPath(firstSamPath);
+
+    if (samFormat == null) {
+      throw new IllegalArgumentException("Cannot find format extension for " + path);
+    }
+
+    AbstractSamSource abstractSamSource = samFormat.createAbstractSamSource(fileSystemWrapper);
+
+    SAMFileHeader header =
+        abstractSamSource.getFileHeader(
+            sparkContext, path, validationStringency, referenceSourcePath);
+    JavaRDD<SAMRecord> reads =
+        abstractSamSource.getReads(
+            sparkContext,
+            path,
+            splitSize,
+            traversalParameters,
+            validationStringency,
+            referenceSourcePath);
+    return new HtsjdkReadsRdd(header, reads);
+  }
+
+  /**
+   * Write reads to a file or files specified by the given path. Write options may be specified to
+   * control the format to write in (BAM/CRAM/SAM, if not clear from the path extension), and the
+   * number of files to write (single vs. multiple).
+   *
+   * @param htsjdkReadsRdd a {@link HtsjdkReadsRdd} containing the header and the reads
+   * @param path the file or directory to write to
+   * @param writeOptions options to control aspects of how to write the reads (e.g. {@link
+   *     ReadsFormatWriteOption} and {@link FileCardinalityWriteOption}
+   * @throws IOException if an IO error occurs while writing
+   */
+  public void write(HtsjdkReadsRdd htsjdkReadsRdd, String path, WriteOption... writeOptions)
```

Comment:
I know the Java APIs uses varargs like this in a few places, but I always find it awkward especially in library APIs. Except in a a few rare cases like String formatting. Perhaps I'm the only one.

Reply:
Thanks for the comments @cmnbroad. I've responded to them below. I don't think any of them should hold up committing this as they can all be addressed separately.

> I'd propose the use of "disq" as a prefix at the top level.

So that would be `DisqReads` etc? That would duplicate the project/package name in the class, which is a bit redundant. I looked at generifying briefly but didn't see a huge gain. We can revisit that though.

> I'd prefer gradle and TestNG over maven and Junit

I'm more familiar with Maven and JUnit, but I don't have an issue with the others.

> Whats the state of 2bit ref support ?

It's not supported. Ideally we'd have 2bit support in htsjdk, then we can use it here.

> Would be good to get a couple of hg38 test files in early on

Agreed

> Are we going to support BCF?

I wasn't planning to, unless there's a lot of demand for it.

> I know the Java APIs uses varargs like this in a few places, but I always find it awkward especially in library APIs.

Do you have a preferred alternative?